src/native/_auth.js

```diff
@@ -132,12 +132,115 @@ const AuthGranted = Struct({
+    enc_key: new t.SYM_KEYBYTES(appKeys.enc_key),
+    sign_pk: new t.SIGN_PUBLICKEYBYTES(appKeys.sign_pk),
+    sign_sk: new t.SIGN_SECRETKEYBYTES(appKeys.sign_sk),
+    enc_pk: new t.ASYM_PUBLICKEYBYTES(appKeys.enc_pk),
+    enc_sk: new t.ASYM_SECRETKEYBYTES(appKeys.enc_sk),
+  });
+}
+
+const makeAccessContInfo = (accessContainer) => {
+  return new AccessContInfo({
+    id: new t.XOR_NAME(accessContainer.id),
+    tag: accessContainer.tag,
+    nonce: new t.SYM_NONCEBYTES(accessContainer.nonce),
+  });
+}
+
+const makeAccessContainerEntry = (accessContainerEntry) => {
+  const contInfoArray = new ContainerInfoArray(accessContainerEntry.length);
+  accessContainerEntry.forEach((entry, index) => {
+    const permissions = new t.PermissionSet({
+      Read: entry.permissions.Read,
+      Insert: entry.permissions.Insert,
+      Update: entry.permissions.Update,
+      Delete: entry.permissions.Delete,
+      ManagePermissions: entry.permissions.ManagePermissions
+    });
+    contInfoArray[index] = new ContainerInfo({
+      name: ref.allocCString(entry.name),
+      mdata_info: helpersForNative.makeMDataInfo(entry.mdata_info),
+      permissions
+    });
+  });
+
+  return new AccessContainerEntry({
+    containers: contInfoArray.buffer,
+    containers_len: contInfoArray.length
+  });
+}
+
+const makeAuthGranted = (authGrantedObj) => {
+  return new AuthGranted({
+    app_keys: makeAppKeys(authGrantedObj.app_keys),
+    access_container: makeAccessContInfo(authGrantedObj.access_container),
+    access_container_entry: makeAccessContainerEntry(authGrantedObj.access_container_entry),
+    bootstrap_config_ptr: new Buffer(authGrantedObj.bootstrap_config),
+    bootstrap_config_len: authGrantedObj.bootstrap_config.length,
+  });
+}
+
+const readAppKeys = (appKeys) => {
+  return {
+    owner_key: new Buffer(appKeys.owner_key),
+    enc_key: new Buffer(appKeys.enc_key),
+    sign_pk: new Buffer(appKeys.sign_pk),
+    sign_sk: new Buffer(appKeys.sign_sk),
+    enc_pk: new Buffer(appKeys.enc_pk),
+    enc_sk: new Buffer(appKeys.enc_sk),
+  };
+}
+
+const readAccessContainer = (accessContainer) => {
+  return {
+    id: new Buffer(accessContainer.id),
+    tag: accessContainer.tag,
+    nonce: new Buffer(accessContainer.nonce),
+  };
+}
+
+const readAccessContainerEntry = (accessContainerEntry) => {
+  let ptr = accessContainerEntry.containers;
+  const len = accessContainerEntry.containers_len;
+  const arrPtr = ref.reinterpret(ptr, ContainerInfo.size * len);
+  const arr = ContainerInfoArray(arrPtr);
+  let containersList = [];
+  for (let i = 0; i < len ; i++) {
+    const currValue = arr[i];
+    containersList.push({
+      name: currValue.name,
+      mdata_info: helpersForNative.makeMDataInfoObj(currValue.mdata_info),
+      permissions: helpersForNative.readPermsSet(currValue.permissions),
+    });
+  }
+  return containersList;
+}
+
+const readAuthGrantedPtr = (authGrantedPtr) => {
```

Comment:
`readAuthGrantedPtr`, what's 'Ptr'?

Reply:
When we receive the `AuthGranted` FFI struct, we actually receive a pointer/ref to it, so `ptr` stands for pointer.